### ISSUE TITLE
gui rpc drop keystroke flags like repeat

### ISSUE THREAD

**Issue by iacore** [QubesOS/qubes-issues#7231]
[How to file a helpful issue](https://www.qubes-os.org/doc/issue-tracking/)

### Qubes OS release
r4.1

### Brief summary

Key repeat is usable in dom0, but the Qubes gui protocol `gui-common` doesn't include flags for keys.

This makes games extremely confused, because they can't differentiate between "pressing a key repeatedly" and "holding down a key".

### Steps to reproduce

Run `xinput test-xi2` in dom0. Hold any key. You will see something like this in stdout:
```
EVENT type 2 (KeyPress)
    device: 3 (3)
    detail: 113
    flags: repeat
```

Run `xinput test-xi2` in a VM other than dom0. Hold any key. The flags are empty.
```
EVENT type 2 (KeyPress)
    device: 3 (3)
    detail: 113
    flags:
```

### Expected behavior

X input events should include flags like `repeat` too.

**Comment by locriacyber** [comment]
In `qubes-gui-protocol.h`, seems like there is no flags here. I could really use some comments.
```
struct msg_keypress {
    uint32_t type;
    uint32_t x;
    uint32_t y;
    uint32_t state;
    uint32_t keycode;
};
```

**Comment by DemiMarie** [comment]
Ouch.  Thanks for reporting this.

@marmarek is this something that can be fixed in an R4.1 update?  On the one hand it does require a protocol change, but on the other hand it seems to be causing actual problems for at least one user.

**Comment by locriacyber** [comment]
Maybe...
- create a new event KEYPRESS2
- make gui rpc server emit both KEYPRESS and KEYPRESS2
- make new rpc client only handle KEYPRESS2

**Comment by locriacyber** [comment]
Mouse scroll wheel event also has `flags`. Has not met problem with this so far.
```
EVENT type 4 (ButtonPress)
    device: 25 (25)
    detail: 5
    flags: emulated
```

**Comment by airelemental** [comment]
When trying online game streaming services (like Stadia), I had to turn off keyboard repeat in dom0 (to force it to send repeated keystrokes individually?) It sounds like this is the cause.

**Comment by locriacyber** [comment]
There are too many `RawKeyPress` events, one for each repeated key event. There should only be one when the key is initially pressed.

**Comment by locriacyber** [comment]
Is there a plan to use TLV encoding for qubes events? That way, we can extend existing events with new fields in the future.

**Comment by DemiMarie** [comment]
@locriacyber Not that I am aware of.  The encoding is deliberately kept as simple as possible to minimize attack surface.  The protocol does support version negotiation, though, and the agent ignores events it is not aware of.

**Comment by locriacyber** [comment]
Ok, I found a partial solution. I turned off keyboard repeat in "Keyboard" application in dom0, now every client VM seem to generate the correct key repeat events themselves.

Still can't turn off emulated events for mouse wheels.

**Comment by locriacyber** [comment]
Also, mouse wheel motion events are never sent to client VM.

**Comment by iacore** [comment]
I'm not using QubesOS anymore, so how do I fix this? I don't think I can.

**Comment by DemiMarie** [comment]
I’m not sure this is reasonably fixable with the current GUI protocol.  The problem is not the quality of the PR or the complexity of the fix, but rather the lack of anyone with sufficient knowledge of X11 to review the changes.  Most who worked on X11 have since moved on to Wayland, so it is very difficult to get help for X11 problems.

@iacore: Which documentation did you use to figure outt how XInput worked?

**Comment by iacore** [comment]
> I’m not sure this is reasonably fixable with the current GUI protocol. 

I agree. Given how xserver works, state synchronization with X11 feels like a meme. Better to patch the xserver to sync the internal states directly.

> @iacore: Which documentation did you use to figure outt how XInput worked?

None.

I ran the following two commands and tried to make the output match.

```
xinput test-xi2
xinput test-xi2 --root
```

**Comment by iacore** [comment]
> Which X server would need to be patched? Agent (guest) side or daemon (host) side?

i forgot what the Qubes OS gui architecture is like. i remembering window focus event and keyboard input event arrive in different order or just don't arrive at all. i fixed the alt+tab to work fairly consistently, but overall using Qubes OS still feels weird.

I forgot a lot about how i did what, so take a look at https://github.com/QubesOS/qubes-gui-daemon/pull/115 and use `xinput test-xi2` to see if the output in a user VM is acceptable for you.

**Comment by DemiMarie** [comment]
I’m pretty sure this has not been fixed.  QubesOS/qubes-gui-daemon#115 has been fixed to no longer rely on raw events, so it ought to be merged if things work.

**Comment by Atrate** [comment]
I can confirm that this is still an issue.

The only real workaround is to pass-through the entire keyboard via USB passthrough, which is heavily not recommended (seen here: https://www.qubes-os.org/doc/device-handling-security/#security-warning-on-usb-input-devices).

As such I feel like it'd be nice if this got a bit of a priority bump, as the current behavior encourages insecure behavior as workarounds (or avoiding workflows that necessitate this, which is not a solution).

**Comment by Atrate** [comment]
I'd like to put a 25 USD bounty on this.

**Comment by Atrate** [comment]
This is still an issue on r4.3.

**Comment by kalidatuna** [comment]
@Atrate — is your **$25 USD bounty** on this issue still available?

I’m interested in taking it, but before I start I want to confirm the payment/acceptance side once rather than create another speculative patch. I’m 17 and based in Georgia, and I use AI-assisted development that I review and test myself. If that is eligible, can you confirm:

1. the $25 bounty is still open and unclaimed;
2. whether payment can be made by PayPal or international bank transfer after acceptance; and
3. whether your acceptance condition is an upstream-merge, or a tested patch that fixes the missing key-repeat flag behavior even if upstream review is still pending?

No payment details need to be posted publicly. I’ll wait for confirmation before implementing.
